Add DFU support for intel_s1000.

SavinayDharmappa opened the pull request:
The bulk of implementation in the current DFU arch is done in the ISR.
    This works well when the Flash device is memory mapped as these writes
    get done comparatively quickly. However, in case of platforms where
    the flash device is sitting on the SPI Bus, this was observed to
    cause an exception. This exception may be because there are multiple
    function calls consuming larger processing time inside the ISR. The
    ISR stack may also end up going deeper and risk stack corruption on
    devices with low RAM. To resolve this, we deferred flash write to a
    worker thread. Also, some handshaking was added to have synchronization
    with the host in accordance with the Host-Device DFU protocol.

SavinayDharmappa committed:
usb: class: dfu: Add a worker thread to flash data from usb to flash

The bulk of implementation in the current DFU arch is done in the ISR.
This works well when the Flash device is memory mapped as these writes
get done comparatively quickly. However, in case of platforms where
the flash device is sitting on the SPI Bus, this was observed to
cause an exception. This exception may be because there are multiple
function calls consuming larger processing time inside the ISR. The
ISR stack may also end up going deeper and risk stack corruption on
devices with low RAM. To resolve this, we deferred flash write to a
worker thread. Also, some handshaking was added to have synchronization
with the host in accordance with the Host-Device DFU protocol.

Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>

SavinayDharmappa commented:
build in ci will fail as this patch has dependency on #11850

nashif commented on boards/xtensa/intel_s1000_crb/Kconfig.defconfig:
```diff
@@ -167,6 +167,12 @@ config GPIO_DW_0
 
 endif
 
+if USB_DFU_CLASS
```
who really needs SPI? USB_DFU or Flash? This should depend on Flash rathe than USB I guess

SavinayDharmappa replied:
ok. I will change it

nashif commented on boards/xtensa/intel_s1000_crb/dts_fixup.h:
```diff
@@ -26,4 +26,6 @@
 #define CONFIG_SPI_NOR_SPI_SLAVE		\
 	DT_SNPS_DESIGNWARE_SPI_E000_SPI_FLASH_0_BASE_ADDRESS
+
+#define DT_FLASH_DEV_NAME DT_SNPS_DESIGNWARE_SPI_E000_SPI_FLASH_0_LABEL
```
This needs to be in a different commit. Where is this being used? I do not see this being used in this PR

SavinayDharmappa replied:
ok.

nashif commented:
> In case of intel_s1000  flash has to be accessed via SPI.
This was resulting in exception to resolve this a worker thread
is added to flash the data received from the USB .
In design ware USB driver DFU protocol  related callbacks are called
from interrupt context, these callback would flash data received
from USB to SPI flash calling design ware SPI driver.
Invoking  SPI “transceive” api in the interrupt was causing
the expectation as this acquires semaphore with K_FOREVER flag.

Commit message is difficult to read and understand, please rephrase and fix it to have proper paragraphs.

nashif commented:
> build in ci will fail as this patch has dependency on #11850

I do not see any dependencies in the PR on #11850

rgundi commented:
> > build in ci will fail as this patch has dependency on #11850
> 
> I do not see any dependencies in the PR on #11850

FLASH_AREA_IMAGE_1_OFFSET is made use of in this PR but it is defined for s1000 only in #11850. However, I am surprised why there's no build error. In fact, i am not sure how this sample gets executed for platforms which don't define FLASH_AREA_IMAGE_1_OFFSET.

SavinayDharmappa committed:
samples: subsys: usb: dfu: Fix build error for intel_s1000

intel_s1000 uses DT_SPI_NOR_DRV_NAME instead of DT_FLASH_DEV_NAME
for device binding of flash device.

Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>

jfischer-no reviewed (commented):
>The bulk of implementation in the current DFU arch is done in the ISR.

It is worse (depending on platform), the usb stack runs in the ISR context. Fixing it requires discussion in larger round. This patch fixes the problem only partially local.

rgundi commented on subsys/usb/class/usb_dfu.c:
```diff
@@ -679,6 +691,25 @@ USBD_CFG_DATA_DEFINE(dfu_mode) struct usb_cfg_data dfu_mode_config = {
 
+static void dfu_work_handler(struct k_work *item)
+{
+
```
Please use ARG_UNUSED for item.

rgundi reviewed (commented):
Minor comment

SavinayDharmappa committed:
samples: subsys: usb: dfu: Add intel_s1000 to platform_whitelist

add intel_s1000 to platform_whitelist

Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>

rgundi reviewed (approved):
LGTM

rgundi commented:
@jfischer-phytec-iot : Yes, this solves only a part of the problem but we consciously decided that way as there are too many dependencies and touching it may have resulted in potential regressions on many boards since USB is all pervasive. However, please note that the modifications done in this PR hold good even if the DFU handler was NOT being called in an ISR since the handler is a "blocking call" which doesn't work for the SPI driver which is interrupt-based. So, this PR is good enough on its own regardless of what happens with the larger USB conundrum.

jfischer-phytec-iot commented on subsys/usb/class/usb_dfu.c:
```diff
@@ -430,13 +434,21 @@ static int dfu_class_handle_req(struct usb_setup_packet *pSetup,
+			dfu_data.state = dfuDNBUSY;
+			dfu_state = dfuDNLOAD_IDLE;
+			dfu_len = pSetup->wLength;
+			if (dfu_len == 0)
```
brackets missed

SavinayDharmappa replied:
done

jfischer-phytec-iot commented on subsys/usb/class/usb_dfu.c:
```diff
@@ -63,6 +63,10 @@ LOG_MODULE_REGISTER(usb_dfu);
 #endif
 
+static struct k_work dfu_work;
+u8_t buf[USB_DFU_MAX_XFER_SIZE];
```
static?

jfischer-phytec-iot commented on subsys/usb/class/usb_dfu.c:
```diff
@@ -63,6 +63,10 @@ LOG_MODULE_REGISTER(usb_dfu);
 
+static struct k_work dfu_work;
+u8_t buf[USB_DFU_MAX_XFER_SIZE];
+u32_t dfu_len, dfu_state;
```
static? I would use dfu_worker_len and dfu_worker_state. Or would it not be better to put it inside dfu_data_t and call them worker_len and worker_state?

SavinayDharmappa replied:
packed all variables related to a worker thread in a structure.

jfischer-phytec-iot commented on subsys/usb/class/usb_dfu.c:
```diff
@@ -679,6 +691,26 @@ USBD_CFG_DATA_DEFINE(dfu_mode) struct usb_cfg_data dfu_mode_config = {
+		dfu_flash_write(buf, dfu_len);
+		break;
+	default:
+		LOG_DBG("OUT of state machine");
```
s/LOG_DBG/LOG_WRN or s/LOG_DBG/LOG_ERR

SavinayDharmappa replied:
done

jfischer-no reviewed (dismissed):
Agree with the workaround (tested on frdm_k64f), just few comments

aurel32 commented:
I look at it quickly and it looks fine to me. Writing through a work queue looks like a good idea, even for memory mapped flashes. I guess the problem is not encountered on SoCs with memory mapped flash, as the flash and thus the CPU (unless it runs from cache) are also stopped.

I have noticed that you haven't moved the upload case to the work queue. Reading my be faster than reading, but I am not sure it's still a good idea to run SPI transfers from the ISR context. I have also a question about the upload feature in link with PR#11027. What is the value of `CONFIG_FLASH_BASE_ADDRESS` on the `intel_s1000`?. I believe that if it is not zero you might encounter the same issue for the upload feature than in the PR. The address passed to `flash_read()` is an absolute address while it expects a relative address.

SavinayDharmappa commented:
> I have noticed that you haven't moved the upload case to the work queue. Reading my be faster than reading, but I am not sure it's still a good idea to run SPI transfers from the ISR context.

Agree. We tried the work queue approach even with the Upload case but that is not functional and requires more work. It is also possible that Upload requires a completely different approach (may be different from work queue). Whatever the approach for Upload is, the work queue approach for Download works well and hence this PR only refers to Download.

> What is the value of `CONFIG_FLASH_BASE_ADDRESS` on the `intel_s1000`?.

It is Zero. So the bug you are fixing in PR#11027 is not applicable to us.

SavinayDharmappa commented:
@jfischer-phytec-iot @aurel32  I have addressed all your comments. Could you please check if there is anything else pending? if not, can this be approved and merged

SavinayDharmappa committed:
boards: xtensa: intel_s1000_crb: set USB_DFU_MAX_XFER_SIZE to 4096

patch set the config USB_DFU_MAX_XFER_SIZE to 4096. This will
reduce the time required for flashing.

Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>

aurel32 commented:
> @jfischer-phytec-iot @aurel32 I have addressed all your comments. Could you please check if there is anything else pending? if not, can this be approved and merged

@SavinayDharmappa Thanks from addressing my comments. The idea of using a worker and the implementation both look fine, however I don't have time right now to give it a try. I therefore do not want to approve this PR, but I am fine if someone else approves it and it get merged.
